File: 2025/01/main.py
def part1():
    dial,zeroes = 50,0
    for line in open("input.txt"):
        dial = (dial + ((-1 if line[0] == "L" else 1) * int(line[1:])) + 10000) % 100
        zeroes += (1 if dial == 0 else 0)
    return zeroes

def part2():
    dial,zeroes = 50,0
    for line in open("input.txt"):
        direction,distance = line[0],int(line[1:])
        if direction == "R":
            zeroes += (dial + distance) // 100
        else:
            zeroes += distance // 100 + (1 if (distance % 100 >= dial > 0) else 0)
        dial = (dial + ((-1 if direction == "L" else 1) * distance) + 10000) % 100
    return zeroes

if __name__ == "__main__":
    print("Part 1:",part1())
    print("Part 2:",part2())

File: 2025/01/main.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

type instruction struct {
	direction byte
	distance  int
}

func parseInput() []instruction {
	data, _ := os.ReadFile("input.txt")
	lines := strings.Split(strings.TrimSpace(string(data)), "\n")

	instructions := make([]instruction, len(lines))
	for i, line := range lines {
		distance, _ := strconv.Atoi(line[1:])
		instructions[i] = instruction{line[0], distance}
	}
	return instructions
}

func updateDial(dial, distance int, direction byte) int {
	if direction == 'L' {
		return (dial - distance%100 + 100) % 100
	}
	return (dial + distance) % 100
}

func part1(instructions []instruction) int {
	dial, zeroes := 50, 0
	for _, inst := range instructions {
		dial = updateDial(dial, inst.distance, inst.direction)
		if dial == 0 {
			zeroes++
		}
	}
	return zeroes
}

func part2(instructions []instruction) int {
	dial, zeroes := 50, 0
	for _, inst := range instructions {
		if inst.direction == 'R' {
			zeroes += (dial + inst.distance) / 100
		} else {
			zeroes += inst.distance / 100
			if inst.distance%100 >= dial && dial > 0 {
				zeroes++
			}
		}
		dial = updateDial(dial, inst.distance, inst.direction)
	}
	return zeroes
}

func main() {
	instructions := parseInput()
	fmt.Println("Part 1:", part1(instructions))
	fmt.Println("Part 2:", part2(instructions))
}
